cylc/uiserver/distributed_spawner.py

```diff
@@ -0,0 +1,203 @@
+import logging
+from subprocess import Popen, PIPE, DEVNULL
+import sys
+from textwrap import dedent
+from typing import List, Tuple
+
+from traitlets import (
+    DottedObjectName,
+    List as TList,
+    Unicode,
+    default,
+)
+from jupyterhub.spawner import Spawner
+
+from cylc.flow import __version__ as CYLC_VERSION
+from cylc.flow.host_select import select_host
+
+
+logger = logging.getLogger(__name__)
+
+
+class DottedObject(DottedObjectName):
+    """Like DottedObjectName, only it actually imports the thing."""
+
+    def validate(self, obj, value):
+        """Import and return bar given the string foo.bar."""
+        package = '.'.join(value.split('.')[0:-1])
+        obj = value.split('.')[-1]
+        try:
+            if package:
+                module = __import__(package, fromlist=[obj])
+                return module.__dict__[obj]
+            else:
+                return __import__(obj)
+        except ImportError:
+            self.error(obj, value)
+
+
+class DistributedSpawner(Spawner):
+    """A simple SSH Spawner with load balancing capability.
+
+    Runs as the user, no elevated privileges required.
+
+    Requires both passphraseless SSH and a shared filesystem between the
+    hub server and all configured hosts.
+    """
+
+    hosts = TList(
+        trait=Unicode(),
+        config=True,
+        help='''
+            List of host names to choose from.
+        '''
+    )
+
+    ranking = Unicode(
+        config=True,
+        help='''
+            Ranking to use for load balancing purposes.
+
+            If unspecified a host is chosen at random.
+
+            These rankings can be used to pick the host with the most available
+            memory or filter out hosts with high server load.
+
+            These rankings are provided in the same format as
+            :cylc:conf`global.cylc[scheduler][run hosts]ranking`.
+        '''
+    )
+
+    ssh_cmd = TList(
+        trait=Unicode(),
+        config=True,
+        help='''
+            The SSH command to use for connecting to the remote hosts.
+
+            E.G: ``['ssh']`` (default)
+        '''
+    )
+
+    get_ip_from_hostname = DottedObject(
+        config=True,
+        help='''
+            Function for obtaining the IP address from a hostname.
+
+            E.G: ``socket.gethostbyname`` (default)
+        '''
+    )
+
+    @default('get_ip_from_hostname')
+    def default_ip_from_hostname_command(self):
+        return 'socket.gethostbyname'
+
+    @default('ssh_cmd')
+    def default_ssh_command(self):
+        return ['ssh']
+
+    def __init__(self, *args, **kwargs):
+        Spawner.__init__(self, *args, **kwargs)
+        self._proc = None
+
+    def choose_host(self):
+        return select_host(self.hosts, self.ranking)[1]
+
+    def get_env(self):
+        return {
+            **Spawner.get_env(self),
+            'CYLC_VERSION': CYLC_VERSION,
+            'JUPYTERHUB_SERVICE_PREFIX': '/user/osanders/'
```

Comment:
Having a brief glance at this with #230, and not marked ready for review but this line will need changing.
Nice looking pr though, happy to review when it is ready.

Reply:
Yeah this was hacked a bit, should be much nicer with Jupyter Server :crossed_fingers:.